src/devices/sound/setapcm.cpp

```diff
@@ -0,0 +1,338 @@
+
+      00-1f: Voice 0
+      Offset Bit               Description
+             fedcba98 76543210 
+      04     xxxxxxxx xxxxxxxx Start position LSB
+      06     xxxxxxxx xxxxxxxx Start position MSB
+      0a     -------- ----xxx- Used but unknown
+             -------- -----x-- See below for NiLe specific? notes
+             -------- -------x Loop enable
+      0c     xxxxxxxx xxxxxxxx Frequency
+      0e     xxxxxxxx xxxxxxxx Loop Start position LSB
+      12     xxxxxxxx xxxxxxxx Loop Start position MSB
+      14     xxxxxxxx xxxxxxxx Loop End position LSB
+      16     xxxxxxxx xxxxxxxx Loop End position MSB
+      18     xxxxxxxx xxxxxxxx End position LSB
+      1a     xxxxxxxx xxxxxxxx End position MSB
+      1c     xxxxxxxx xxxxxxxx Right Volume
+      1e     xxxxxxxx xxxxxxxx Left Volume
+
+      20-3f: Voice 1
+      ...
+	  e0-ff: Voice 7
+
+      100: Keyon/off, Bit 0-7 means Voice 0-7
+      110: Used but unknown
+
+      below for 16 voice configurations:
+      100-11f: Voice 8
+      120-13f: Voice 9
+      ...
+      1e0-1ff: Voice 15
+
+      200: Keyon/off, Bit 0-15 means Voice 0-15
+      210: Used but unknown
+
+      Other registers are unknown/unused
+
+      TODO:
+      - Verify loop and flag bit behavior from real hardware
+
+************************************/
+
+#include "emu.h"
+#include "setapcm.h"
+
+DEFINE_DEVICE_TYPE(NILE_SOUND,   nile_sound_device,   "nile_sound",   "Seta ST-0026 NiLe (Sound)")
+DEFINE_DEVICE_TYPE(ST0032_SOUND, st0032_sound_device, "st0032_sound", "Seta ST-0032 (Sound)")
+
+setapcm_device::setapcm_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_sound_interface(mconfig, *this)
+	, device_rom_interface(mconfig, *this)
+	, m_stream(nullptr)
+	, m_ctrl(0)
+{
+}
+
+nile_sound_device::nile_sound_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: setapcm_device(mconfig, NILE_SOUND, tag, owner, clock)
+{
+	m_max_voices = 8;
+}
+
+st0032_sound_device::st0032_sound_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: setapcm_device(mconfig, ST0032_SOUND, tag, owner, clock)
+{
+	m_max_voices = 16;
+}
+
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void setapcm_device::device_start()
+{
+	m_stream = stream_alloc(0, 2, clock()); // TODO: verify from real hardware
+
+	save_item(STRUCT_MEMBER(m_voice, m_start));
+	save_item(STRUCT_MEMBER(m_voice, m_flags));
+	save_item(STRUCT_MEMBER(m_voice, m_freq));
+	save_item(STRUCT_MEMBER(m_voice, m_lpstart));
+	save_item(STRUCT_MEMBER(m_voice, m_lpend));
+	save_item(STRUCT_MEMBER(m_voice, m_end));
+	save_item(STRUCT_MEMBER(m_voice, m_vol_r));
+	save_item(STRUCT_MEMBER(m_voice, m_vol_l));
+	save_item(STRUCT_MEMBER(m_voice, m_pos));
+	save_item(STRUCT_MEMBER(m_voice, m_frac));
```

Comment:
Do you think that having these aliased in `m_sound_regs` but only updated on register writes is good design?  There are two copies of the value, one in `m_voice` and one in `m_sound_regs`.  The one in `m_voice` is update on writes and also when updating the output, but the one in `m_sound_regs` is only updated on write.  This could be confusing to someone viewing `m_sound_regs` in the debugger.